### PR TITLE
Add Primitive Definitions

### PR DESCRIPTION
**Pull request by dan-fritchman**
* Adds "message literal" definitions for primitive elements, in protobuf-text format. 
* Adds a (python) script which generates those definitions 
* (TBC) updates netlisting to use said primitives, rather than relying on HDL21.

### CODE REVIEW

File: bindings/python/vlsir/primitives.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+from pathlib import Path
+from .circuit_pb2 import Package
+
+# Get the definition-file from this file's directory
+here = Path(__file__).parent.absolute()
+prims = here / "vlsir.primitives.pb.txt"
+
+# And parse it into a `circuit.Package`
+pkg = Package()
+(open(prims, "r").read(), pkg)
```
Comment (aviralpandey) [inline bindings/python/vlsir/primitives.py:16]: Don't understand how this Package creation method works

**Review by aviralpandey — APPROVED** [review]
Looks good to me. 

A few high questions: 

1) Are the parameters all required, or are they optional? 
2) Do we want to split-up the proto versions of various current/voltage sources? E.g. make a VPulse different from a VDC different from a VAC different from a VSin?

File: bindings/python/vlsir/primitives.py (hunk 1 above)
Comment (dan-fritchman) [reply bindings/python/vlsir/primitives.py:16]: Yeah you're right, it doesn't work. 
Line 16 is supposed to say: 

```python 
text_format.Parse(open(prims, "r").read(), pkg)
```

And this begs for a better unit-test than the existing one: 

```python
def test_primitives1():
    from vlsir.primitives import pkg
    assert isinstance(pkg, vlsir.circuit.Package)
```

Which passes even without parsing the content.

**Comment by dan-fritchman** [comment]
**Are the parameters all required, or are they optional?**

A fine question, complicated in some contexts. 
As far as the `Message`-class-level goes, all are optional. Making anything required would be done by us during netlisting (or by SPICE on the resultant netlist). SPICE will inject some defaults of its own, which I think this works in our favor in many cases. You could imagine, for example, `resistor.tc2`, a second-order tempco of a resistor, being a parameter that exists, takes on a default value if not specified, and we really don't want or care to look up any given SPICE's default value. 

There's also the topic of parameters that spice-level stuff has, and we don't necessarily want to explicitly spell out. Adding some commentary from Slack:

---

One important question is the parameter-space for each.
For `mos`, I think the diversity of models and their parameters is pretty hopeless to explicitly spell out. There is a long doc-string to this effect for their primitive, which explicitly declares a single parameter: the model-name. (Any other params then need to get “passed through” it.)

This also implies that one can pass such undeclared parameters into these ExternalModules, or at least into some of them. (For example some MOSes will want, say, `nfin` to be a valid parameter, some will not.) The new statement here is: *such parameter-instances are supported*. Perhaps instead they should only be OK some of the time, i.e. for primitives or via another option.

---

**Do we want to split-up the proto versions of various current/voltage sources? E.g. make a VPulse different from a VDC different from a VAC different from a VSin?**

Certainly a fine topic for debate, and ties into the parameter-handling above. Personally I kinda like something like the Spectre `vsource`, a single thing that has essentially the union of the parameter-spaces for dc, pulse, sine, etc. Although it does make for some more complicated statements of which parameter does what when.

**Comment by dan-fritchman** [comment]
Latest commit 3c287fc does break out voltage-sources into discrete `vdc`, `vsin`, `vpulse`. (Note `vpwl` would be the next logical choice, but would allow for some more elaborate parameter-types in the schema, e.g. lists.) 

The parameter-space between each of those is so disjoint, it makes sense for both users and the vlsir codebase to break them up. 

This commit (hopefully) also has all of the subtleties for parametrizing all of these in out favorite-simulators' syntaxes.

**Pull request merged**